### PR TITLE
jprofiler-agent: init at 13.0.2

### PR DESCRIPTION
**Pull request by catap**
###### Description of changes

I've also heavy refactored jprofiler port that makes possible to install the last version for macOS.

An agent is very small library which should be integrated into target application to allow remote debugging.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by catap** [comment]
@azahi may I ask you to review this one?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -16702,7 +16702,10 @@ with pkgs;
 
   java-language-server = callPackage ../development/tools/java/java-language-server { };
 
-  jprofiler = callPackage ../development/tools/java/jprofiler {
+  jprofiler-agent = callPackage ../development/tools/java/jprofiler/agent.nix {
+    jdk = jdk11;
+  };
+  jprofiler = callPackage ../development/tools/java/jprofiler/ui.nix {
     jdk = jdk11;
   };
```
Comment (azahi) [inline pkgs/top-level/all-packages.nix:16710]:
```suggestion
  jprofiler = callPackage ../development/tools/java/jprofiler/ui.nix {
    jdk = jdk11;
  };
  jprofiler-agent = jprofiler.override {
    componenet = "agent";
  };
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/java/jprofiler/common.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+in stdenv.mkDerivation (finalAttrs: rec {
+  pname = "jprofiler";
+  version = options.version;
+
+  meta = with lib; {
+    description = "JProfiler's ${component} helps you resolve performance bottlenecks";
+    longDescription = ''
+      JProfiler's ${component} helps you resolve performance bottlenecks,
+      pin down memory leaks and understand threading issues.
+    '';
```
[the remark this answers is not in the document]
Comment (catap) [reply pkgs/development/tools/java/jprofiler/common.nix:30]: @azahi agent brings a very small (~7mb) set of libraries which can be used to add to existed application to allow remote profiling. From another hand the UI is full graphical application to use this agent.

The idea is:
 - install jprofiler (=UI) to developer machine;
 - install services with agent library;
 - connect from UI to agent by network when needed.

File: pkgs/development/tools/java/jprofiler/test.nix [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{ runCommand, jprofiler }:
+
+runCommand "jprofiler-test-run" {
+  nativeBuildInputs = [ jprofiler ];
+} ''
+  diff -U3 --color=auto <((jpcontroller -n -f /dev/null 2>&1 || true) | head -n 1 ) <(echo 'No profiled JVMs found.')
+  touch $out
```
[the remark this answers is not in the document]
Comment (catap) [reply pkgs/development/tools/java/jprofiler/test.nix:7]: This way of running `jpcontroller` forces it to load platform-related libraries via JNI and confirm that it is relinked on the right way.

To be honest this test allows to discover that:
 - on Linux is should be relinked against X11;
 - on macOS it should be done with `dontStrip = true` otherwise a signature won't pass and it fails.

The second one is tricky because UI is started and it seems working, but the real attempt to use it leads to fail.

File: pkgs/development/tools/java/jprofiler/common.nix [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    cp -r . $out
+
+    rm -f $out/bin/updater || true
+    rm -rf $out/bin/updater.app || true
+    rm -rf $out/bin/linux-ppc* || true
+    rm -rf $out/bin/linux-armhf || true
+    rm -rf $out/bin/linux_musl* || true
```
[the remark this answers is not in the document]
Comment (catap) [reply pkgs/development/tools/java/jprofiler/common.nix:67]: @azahi do you have an idea how to do it?

[the remark this answers is not in the document]
Comment (azahi) [reply pkgs/development/tools/java/jprofiler/common.nix:67]: Create an attrset where the name is Nixpkgs-compatible platform name, and the value is name of the directory in `./bin`. With a simple `optionalString` delete other directories or don't copy them altogether depending on the value of `stdenv` platform. You can probably reuse an attrset where you declare all available tarballs.

[the remark this answers is not in the document]
Comment (catap) [reply pkgs/development/tools/java/jprofiler/common.nix:66]: @azahi that makes things very complicated and adds headache for updating. From another hand the real reason to remove it is fixing build on linux which tries to relink musl's compiled libraries and fails.

=> I'd love to remove all this removing some parts and simple add required dependencies

**Comment by catap** [comment]
@azahi to make review easy I haven't rebase anything. I've added two commits: (1) addresses all remarks, and (2) moving files as suggest.

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 5]
```diff
@@ -1,45 +1,45 @@
...
-    description = "JProfiler's intuitive UI helps you resolve performance bottlenecks";
+    description = "JProfiler's ${component} helps you resolve performance bottlenecks";
     longDescription = ''
-      JProfiler's intuitive UI helps you resolve performance bottlenecks,
+      JProfiler's ${component} helps you resolve performance bottlenecks,
       pin down memory leaks and understand threading issues.
     '';
     homepage = "https://www.ej-technologies.com/products/jprofiler/overview.html";
     license = licenses.unfree;
+    platforms = with lib.platforms; linux ++ darwin;
```
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:32]: This is the default.
```suggestion
```

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 6]
```diff
@@ -1,45 +1,45 @@
...
+, callPackage
 , jdk
+, libX11
+, component ? "ui"
 }:
 
 let
-  inherit (stdenv.hostPlatform) system;
-  pname = "jprofiler";
+  options = import ./version.nix;
```
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:16]: Maybe you should name it something like this for clarity.
```suggestion
  versions = import ./versions.nix;
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 6 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:16]: I not sure that `versions` is better naming. But maybe it is.

**Comment by catap** [comment]
@azahi I've added one more commit. After you're OK with this changes I'd love to squash everything into one commit.

**Comment by azahi** [comment]
Fails for me with:
```
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/azahi/.cache/nixpkgs-review/pr-179963/build.nix
error: builder for '/nix/store/1lcqsmsvg5nmwmn3k97v7z4lb0xk43aw-jprofiler-13.0.2.drv' failed with exit code 1;
       last 10 log lines:
       >     libc.musl-x86_64.so.1 -> not found!
       > setting interpreter of /nix/store/slqps7r122vj0hkbn26bxp6ydfc48qch-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       > searching for dependencies of /nix/store/slqps7r122vj0hkbn26bxp6ydfc48qch-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       >     libc.musl-x86_64.so.1 -> not found!
       > auto-patchelf: 3 dependencies could not be satisfied
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/slqps7r122vj0hkbn26bxp6ydfc48qch-jprofiler-13.0.2/bin/linux_musl-x64/libjphelper.so
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/slqps7r122vj0hkbn26bxp6ydfc48qch-jprofiler-13.0.2/bin/linux_musl-x64/libjprofilerti.so
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/slqps7r122vj0hkbn26bxp6ydfc48qch-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       > auto-patchelf failed to find all the required dependencies.
       > Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
       For full logs, run 'nix log /nix/store/1lcqsmsvg5nmwmn3k97v7z4lb0xk43aw-jprofiler-13.0.2.drv'.
error: builder for '/nix/store/q0ww717xgdjzlhck778093gwpaig9bmr-jprofiler-13.0.2.drv' failed with exit code 1;
       last 10 log lines:
       >     libc.musl-x86_64.so.1 -> not found!
       > setting interpreter of /nix/store/rr1yxkc81m533jcc74crdxqbzha29v9v-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       > searching for dependencies of /nix/store/rr1yxkc81m533jcc74crdxqbzha29v9v-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       >     libc.musl-x86_64.so.1 -> not found!
       > auto-patchelf: 3 dependencies could not be satisfied
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/rr1yxkc81m533jcc74crdxqbzha29v9v-jprofiler-13.0.2/bin/linux_musl-x64/libjphelper.so
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/rr1yxkc81m533jcc74crdxqbzha29v9v-jprofiler-13.0.2/bin/linux_musl-x64/libjprofilerti.so
       > error: auto-patchelf could not satisfy dependency libc.musl-x86_64.so.1 wanted by /nix/store/rr1yxkc81m533jcc74crdxqbzha29v9v-jprofiler-13.0.2/bin/linux_musl-x64/.jpnetcat-wrapped
       > auto-patchelf failed to find all the required dependencies.
       > Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
       For full logs, run 'nix log /nix/store/q0ww717xgdjzlhck778093gwpaig9bmr-jprofiler-13.0.2.drv'.
error: 2 dependencies of derivation '/nix/store/ry3ac8vn3zmh7sfd1gfacnyfqgdj9bpa-review-shell.drv' failed to build
```

**Comment by catap** [comment]
@azahi can you re-review one more time?

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 179963` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>jprofiler</li>
    <li>jprofiler-agent</li>
  </ul>
</details>

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 179963` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>jprofiler</li>
    <li>jprofiler-agent</li>
  </ul>
</details>

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 7]
```diff
@@ -1,45 +1,45 @@
...
 , jdk
+, libX11
+, musl
+, component ? "ui"
 }:
 
 let
-  inherit (stdenv.hostPlatform) system;
-  pname = "jprofiler";
+  versions = import ./version.nix;
```
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:17]: Since you've changed the name, you should also change the filename.
```suggestion
  versions = import ./versions.nix;
```

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 8]
```diff
@@ -1,45 +1,45 @@
...
 
-  # 11.1.4 is the last version which can be unpacked by undmg
-  # See: https://github.com/matthewbauer/undmg/issues/9
-  version = if stdenv.isLinux then "13.0.2" else "11.1.4";
   nameApp = "JProfiler";
 
+in stdenv.mkDerivation (finalAttrs: rec {
```
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:21]: Pretty sure this can be removed now.
```suggestion
in stdenv.mkDerivation rec {
```

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 9]
```diff
@@ -51,49 +51,32 @@ let
...
+    runHook postInstall
+  '';
 
-    meta = meta // { platforms = lib.platforms.darwin; };
+  passthru.tests.run = callPackage ./test.nix {
+    jprofiler = finalAttrs.finalPackage;
   };
-in
-if stdenv.isDarwin then darwin else linux
+})
```
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:82]:
```suggestion
}
```

**Review by azahi — CHANGES_REQUESTED** [review]
Just a couple of small bits we've missed.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 7 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix]: I guess I've found much better naming: artefacts.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 8 above)
Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix:21]: There's no need to use it there. Take a look at [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/faircamp/default.nix#L33).

Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:21]: @azahi I feel that I've missed how can I define used packet (`jprofiler` or `jprofiler-ui`) inside test to use the right path.

Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix:21]: Oh, good point. Okay then.

Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:21]: @azahi let me squash everything into one commit when

**Review by azahi — APPROVED** [review]

**Comment by catap** [comment]
@azahi I've squashed everything to one commit.

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 10]
```diff
@@ -51,49 +51,32 @@ let
...
-  };
 
-  darwin = stdenv.mkDerivation {
-    inherit pname version src;
+      ${lib.optionalString stdenv.isDarwin ''
+        mkdir -p $out/Applications
+        ln -s $out/bin/${pname}.app $out/Applications/${nameApp}.app
+        ln -s ${jdk.home} $out/.install4j/jre
+      ''}
+    ''}
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:74]: Please use ```lib.optionalString bool ''
  ...
``
```.

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 11]
```diff
@@ -51,49 +51,32 @@ let
...
-      rm -f $out/bin/updater
-      rm -rf $out/bin/linux-ppc*
-      rm -rf $out/bin/linux-armhf
-      rm -rf $out/bin/linux-musl*
+    cp -r . $out
 
-      for f in $(find $out/bin -type f -executable); do
-        wrapProgram $f --set JAVA_HOME "${jdk.home}"
-      done
+    for f in $(find $out/bin $out/.install4j -type f -executable); do
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:61]: What is ``.install4j`` doing in ``$out/``?

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 12]
```diff
@@ -51,49 +51,32 @@ let
...
-      runHook preInstall
-      cp -r . $out
+  installPhase = ''
+    runHook preInstall
 
-      rm -f $out/bin/updater
-      rm -rf $out/bin/linux-ppc*
-      rm -rf $out/bin/linux-armhf
-      rm -rf $out/bin/linux-musl*
+    cp -r . $out
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:59]:
```suggestion
    mv -r . $out
```

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 13]
```diff
@@ -51,49 +51,32 @@ let
     categories = [ "Development" ];
   };
 
-  linux = stdenv.mkDerivation {
-    inherit pname version src desktopItems;
-
-    nativeBuildInputs = [ makeWrapper copyDesktopItems ];
+  dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:54]:
```suggestion
  dontStrip = stdenv.isDarwin;
```

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 14]
```diff
@@ -1,45 +1,45 @@
...
   };
 
-  src = if stdenv.isLinux then fetchurl {
-    url = "https://download-gcdn.ej-technologies.com/jprofiler/jprofiler_linux_${lib.replaceStrings ["."] ["_"]  version}.tar.gz";
-    sha256 = "sha256-x9I7l2ctquCqUymtlQpFXE6+u0Yg773qE6MvAxvCaEE=";
-  } else fetchurl {
-    url = "https://download-gcdn.ej-technologies.com/jprofiler/jprofiler_macos_${lib.replaceStrings ["."] ["_"]  version}.dmg";
-    sha256 = "sha256-WDMGrDsMdY1//WMHgr+/YKSxHWt6A1dD1Pd/MuDOaz8=";
-  };
+  src = fetchurl artefacts.${component}.${stdenv.hostPlatform.system};
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:36]: Missing ``or throw unsupported system``.

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 15]
```diff
@@ -1,45 +1,45 @@
...
+in stdenv.mkDerivation (finalAttrs: rec {
+  pname = "jprofiler";
+  version = artefacts.version;
+
   meta = with lib; {
-    description = "JProfiler's intuitive UI helps you resolve performance bottlenecks";
+    description = "JProfiler's ${component} helps you resolve performance bottlenecks";
     longDescription = ''
-      JProfiler's intuitive UI helps you resolve performance bottlenecks,
+      JProfiler's ${component} helps you resolve performance bottlenecks,
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:28]: I don't think we should substitute the descriptions.

File: pkgs/development/tools/java/jprofiler/artefacts.nix [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    };
+    aarch64-darwin = x86_64-darwin;
+
+    x86_64-linux = {
+      url = "https://download-gcdn.ej-technologies.com/jprofiler/jprofiler_linux_${downloadVersion}.tar.gz";
+      hash = "sha256-x9I7l2ctquCqUymtlQpFXE6+u0Yg773qE6MvAxvCaEE=";
+    };
+    i686-linux = x86_64-linux;
+    aarch64-linux = x86_64-linux;
+    armhf-linux = x86_64-linux;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/artefacts.nix:37]: Instead of aliasing systems here we should just use stdenv.isLinux and stdenv.isDarwin.

With this the file should also be inlined into the default.nix.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 11 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:61]: @SuperSandro2000 `.install4j` is https://www.ej-technologies.com/products/install4j/overview.html and it is used as part of jprofiler.

To be honest I can't answer to your question better than "this is a part of distributive"

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 10 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:74]: @SuperSandro2000 I don't get your comment. Sorry, can you rephrase it?

File: pkgs/development/tools/java/jprofiler/artefacts.nix (hunk 16 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/artefacts.nix:37]: @SuperSandro2000 I don't think that `stdenv.isLinux` is good idea because agent has differnet linux distirbutives

**Comment by catap** [comment]
@azahi / @SuperSandro2000 may I ask you to re-review it one more time?

Thanks!

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 11 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/java/jprofiler/default.nix:105]: Shouldn't that be somewhere like $out/share/jprofiler/ instead?

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 17]
```diff
@@ -1,22 +1,70 @@
...
 , makeWrapper
 , makeDesktopItem
 , copyDesktopItems
-, undmg
+, autoPatchelfHook
+, callPackage
 , jdk
+, libX11
+, musl
+, component ? "ui"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:13]: We should add an assert that this is only ui or agent.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 11 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:105]: @SuperSandro2000 nope, and releative paths are hardcoded:
```sh
√ nixpkgs % ls -la /nix/store/kmn9s77swl907mhkbdch2ysxijprl6aa-jprofiler-13.0.2/
total 72
dr-xr-xr-x    15 root  wheel     480 Jan  1  1970 .
drwxrwxr-t  2297 root  nixbld  73504 Aug  1 23:46 ..
dr-xr-xr-x    48 root  wheel    1536 Jan  1  1970 .install4j
dr-xr-xr-x     3 root  wheel      96 Jan  1  1970 Applications
-r--r--r--     1 root  wheel    2301 Jan  1  1970 README.html
dr-xr-xr-x    12 root  wheel     384 Jan  1  1970 api
dr-xr-xr-x    30 root  wheel     960 Jan  1  1970 bin
dr-xr-xr-x     5 root  wheel     160 Jan  1  1970 config
dr-xr-xr-x     7 root  wheel     224 Jan  1  1970 demo
dr-xr-xr-x     8 root  wheel     256 Jan  1  1970 integrations
dr-xr-xr-x    60 root  wheel    1920 Jan  1  1970 lib
-r--r--r--     1 root  wheel   12615 Jan  1  1970 license.html
-r--r--r--     1 root  wheel   11414 Jan  1  1970 license.txt
-r--r--r--     1 root  wheel      24 Jan  1  1970 release
dr-xr-xr-x     4 root  wheel     128 Jan  1  1970 share
√ nixpkgs %  
```

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 17 above)
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:13]: done

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 11 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/java/jprofiler/default.nix:107]: oh, no. So many none FHS paths. Can we move the entire program to somewhere else? Then the relative paths are also correct again

Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:107]: @SuperSandro2000 it is quite difficult and requires some binary patching which brokes signatures on macOS... To be honest, I prefer to not do it without very good reason.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/java/jprofiler/default.nix:120]: something with formatting got messed up but essentially we should optionally append a multiline string instead of nesting it into variable.

File: pkgs/development/tools/java/jprofiler/default.nix (hunk 11 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/java/jprofiler/default.nix:107]: We can't just move everything as is to somewhere else? Bummer

Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:107]: @SuperSandro2000 I can move everything into `$out/opt/jprofiler` and that makes moving `$out/bin/jporfiler` into `$out/opt/jprofiler/bin/jprofiler`.

Comment (SuperSandro2000) [reply pkgs/development/tools/java/jprofiler/default.nix:107]: That would be better. Then add a symlink or wrapper to $out/bin.

Generally packages that are used by end users should not fill / with all kinds of none FHS files.

File: pkgs/development/tools/java/jprofiler/default.nix [hunk 18]
```diff
@@ -51,49 +97,31 @@ let
...
-    meta = meta // { platforms = lib.platforms.linux; };
-  };
 
-  darwin = stdenv.mkDerivation {
-    inherit pname version src;
+      '' + lib.optionalString stdenv.isDarwin ''
+        mkdir -p $out/Applications
+        ln -s $out/bin/${pname}.app $out/Applications/${nameApp}.app
+        ln -s ${jdk.home} $out/.install4j/jre
+      '';
```
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:119]:
```suggestion
      '' + ''
```

**Comment by SuperSandro2000** [comment]
Something is still off with the syntax. Please check that local.
